controllers/dnspolicy_status_updater.go

```diff
@@ -0,0 +1,121 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+
+	"github.com/samber/lo"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+
+	kuadrantdnsv1alpha1 "github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/kuadrant"
+)
+
+func NewDNSPolicyStatusUpdater(client *dynamic.DynamicClient) *DNSPolicyStatusUpdater {
+	return &DNSPolicyStatusUpdater{client: client}
+}
+
+type DNSPolicyStatusUpdater struct {
+	client *dynamic.DynamicClient
+}
+
+func (r *DNSPolicyStatusUpdater) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.updateStatus,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &kuadrantv1alpha1.DNSPolicyGroupKind},
+			{Kind: &DNSRecordGroupKind},
+		},
+	}
+}
+
+func (r *DNSPolicyStatusUpdater) updateStatus(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("DNSPolicyStatusUpdater")
+
+	policyTypeFilterFunc := dnsPolicyTypeFilterFunc()
+	policyAcceptedFunc := dnsPolicyAcceptedStatusFunc(state)
+	policyErrorFunc := dnsPolicyErrorFunc(state)
+
+	policies := lo.FilterMap(topology.Policies().Items(), policyTypeFilterFunc)
+
+	logger.V(1).Info("updating dns policy statuses", "policies", len(policies))
+
+	for _, policy := range policies {
```

Comment:
how do we end up only with policies that target a DNS record under reconciliation? and should there be more than one policy?

Reply:
You don't really, this iterates around all polices ensuring there status is correct. Majority of times this will do nothing, but if any changes are detected in the topology related to a policy, records removed or in a different state etc.., the status will be updated accordingly.

Reply:
I understand now, thanks! This sotw thing is weird